### ISSUE TITLE
Use Function.bind() or lang.hitch() equivalent?

### ISSUE THREAD

**Issue by wkeese** [ibm-js/sdk#4]
As an example, see the preCreate() method in the current version of deliteful/ViewStack.js:

``` js
preCreate: function () {
    ...
    this.on("DOMNodeInserted", this._setChildrenVisibility.bind(this));
},
```

The `bind()` call for `_setChildrenVisibility` precludes the app from attaching to the `_setChildrenVisibility()` method, for example:

``` js
require(["deliteful/ViewStack", "dcl/advise"], function(ViewStack, advise){
    var vs = new ViewStack();
    advise.after(vs, "_setChildrenVisibility", function(){
        console.log("function called");
    });
});
```

This is because `bind()` does eager binding, rather than lazy binding.

The alternative approaches [when setting up binding during the widget creation lifecycle] are to:

(1) use dojo/_base/lang.hitch() or a similar method from our own utility library

``` js
preCreate: function () {
    ...
    this.on("DOMNodeInserted", lang.hitch(this, "_setChildrenVisibility");
},
```

(Note that we can make our own utility library if we do not want to introduce a dojo dependency.)

(2) hardcode a closure

``` js
preCreate: function () {
    ...
    var self = this;
    this.on("DOMNodeInserted",  function(){ self._setChildrenVisibility(); })
},
```

Note that `Function.bind()` can still be used in other cases, after widget creation has finished.

I like the idea of just using `bind()` but I'm starting to think it's too restrictive to our users.

cc @pruzand, @cjolif

**Comment by asudoh** [comment]
Unless we want to let application developers set an advice to a prototype function, we may set the bound function to an instance property and let application developers set an advice to such instance property, like:

``` javascript
preCreate: function () {
    ...
    this.on("DOMNodeInserted", this._boundSetChildrenVisibility = this._setChildrenVisibility.bind(this));
},

...

require(["deliteful/ViewStack", "dcl/advise"], function(ViewStack, advise){
    var vs = new ViewStack();
    advise.after(vs, "_boundSetChildrenVisibility", function(){
        console.log("function called");
    });
});
```

Whether we want to do that may depend on whether we foresee many use cases setting an advice to something like `_setChildrenVisibility()`. If we foresee many use cases of that, I'd prefer 2).

**Comment by wkeese** [comment]
Unfortunately I don't think your this._boundSetChildrenVisibility will work; it's still doing eager binding.

I feel like dojo users did a lot of dojo.connect() calls for the widgets although I forget for what exactly.

**Comment by asudoh** [comment]
Right my `_boundSetChildrenVisibility()` idea was premature.

Most of non-event-listener `dojo.connect()` usages I can see in legacy Dojo projects I've involved in are custom event alternatives (`dojo.connect(ob, "onCustomEvent", "customEventHandler");`) where `ob` can (directly) call `onCustomEvent()` instead of creating an (actual) custom event and dispatching it.

Not sure if there are any applications that use `dojo.connect()` in AOP context extensibly.

**Comment by cjolif** [comment]
At least when the method is not publicly documented (like the one you are referring to) I think we should not care about it and just use `bind`. For documented methods that's indeed a bit different. For documented methods we might want to use the closure solution (2) for now with a `self` variable.

**Comment by wkeese** [comment]
Sounds like a plan, don't forget to add to https://github.com/ibm-js/sdk/blob/master/GUIDELINES.md.

**Comment by cjolif** [comment]
closed in 5236d5e75200aa3c615821b80a0028be5df45149
